### PR TITLE
client: correctly limit the number of outstanding requests

### PR DESCRIPTION
**Pull request by matklad**
Before, we used `TooManyOutstandingRequests` to signal this condition. The problem with that is that by the time we get to invoking user's callback, its already too late --- the code might have blown up with an assertion in `get_message`.

To fix this, check the number of outstanding requests in `get_message` instead. This also allows us to remove the error condition from the request callback altogether.

### CODE REVIEW

File: src/clients/c/tb_client/context.zig
Comment (matklad) [inline src/clients/c/tb_client/context.zig:182]: I _think_ there was a subtle bug  here --- there's also `.register` request, which _won't_ be counted by this counter. New impl should be fine (but `.register` is an extra gotcha for the users to be aware of).

File: src/clients/c/tb_client/context.zig
Comment (matklad) [inline src/clients/c/tb_client/context.zig:303]: This is subtle. In  the `Client.zig`, I actaully allow the users to retain messages beyond request-response itself. Eg, the user can do

```
const message = client.get_message();
some_global_var.message = message.ref();
client.request(message...)
```

So a request might become available either when a reply arrives, or when the user drops their outstanding reference. If the latter case happened here, that would have been a lost wakeup. 

However, tb_client fully controlls message and it doesn't retain references elsewher.

File: src/vsr/client.zig
Comment (matklad) [inline src/vsr/client.zig]: Renamed just to have the CI yell at me if I missed a callsite, will rename back before merge.

File: src/vsr/client.zig
Comment (sentientwaffle) [inline src/vsr/client.zig:27]: :tada:

**Review by sentientwaffle — COMMENTED** [review]
LGTM once the temporary rename is fixed.

**Comment by matklad** [comment]
Adjusted the node client as well. 

For all other clients, there's no other user-visible changes, because they are based on C client, which already:

- emits too-much-concurrency synchroneously
- transparentl retries failed requests

For node, the situation is different --- submitting request can now throw a new error, and request completion callback now always gets undefined as an error. I didn't change the signature of callbacks, though, because I think we are rewriting this to be based on the C client anyway?

**Comment by matklad** [comment]
Ohhh, I can now actually kill the whole retry thing in our C client!

**Review by batiati — APPROVED** [review]
It was nice to get rid of the `retry` list, now the boundary between threads is much more simple to understand: application --> submitted at one side and submitted --> request at the other side.

**Pull request merged**
